README
    This is used to secure the cookies. Encryption keys and message
    authentication keys are derived from this using one-way functions.
    Changing it will invalidate all sessions.

    Number of seconds for which the session may be considered valid. If
    cookie_duration is not set, this is used instead to expire the session
    after a period of time, regardless of the length of the browser
    session. It is unset by default, meaning that sessions expiration is
    not capped.

SYNOPSIS

      # In Dancer 2 config.yml file
    
      session: Cookie
      engines:
        session:
          Cookie:
            secret_key: your secret passphrase
            default_duration: 604800

DESCRIPTION

    This module implements a session factory for Dancer 2 that stores
    session state within a browser cookie. Features include:

      * Data serialization and compression using Sereal

      * Data encryption using AES with a unique derived key per cookie

      * Enforced expiration timestamp (independent of cookie expiration)

      * Cookie integrity protected with a message authentication code (MAC)

    See Session::Storage::Secure for implementation details and important
    security caveats.

SEE ALSO

    CPAN modules providing cookie session storage (possibly for other
    frameworks):

      * Dancer::Session::Cookie -- Dancer 1 equivalent to this module

      * Catalyst::Plugin::CookiedSession -- encryption only

      * HTTP::CryptoCookie -- encryption only

      * Mojolicious::Sessions -- MAC only

      * Plack::Middleware::Session::Cookie -- MAC only

      * Plack::Middleware::Session::SerializedCookie -- really just a
      framework and you provide the guts with callbacks

      * Dancer2::Core::Role::SessionFactory -- documentation of the base
      package, some more attributes to configure the cookie

POD ERRORS

    Hey! The above document had some coding errors, which are explained
    below:

    Around line 1:

      Unknown directive: =attr

    Around line 7:

      Unknown directive: =attr



Changes
Revision history for Dancer2-Session-Cookie

{{$NEXT}}
  [ API CHANGES ]

  [ BUG FIXES ]

  [ DOCUMENTATION ]

  [ ENHANCEMENTS ]

  [ NEW FEATURES ]

  [ STATISTICS ]

0.009 2018-04-25
  - Dependency upped for Session::Storage::Secure. (GH#19, manwar)

  [ STATISTICS ]
    - code churn: 6 files changed, 40 insertions(+), 21 deletions(-)

0.008 2016-11-04
  - Up Dancer2 dependency to 0.165, as 0.164 was a broken one. (GH#13)

  [ DOCUMENTATION ]
    - Clarify where 'cookie_duration' is set. (GH#14)

  [ ENHANCEMENTS ]
    - Add 'with_request_address' option to configuration. (GH#1)

0.007 2016-08-15
  [ BUG FIXES ]
    - Remove context usage (GH#9, Sawyer X)

  [ CHANGED ]
    - Link to base package added in POD (GH#11, Moritz Grosch)
    - Simplify dzil config (Peter Mottram)
    - Rewrite tests to use Plack::Test (Peter Mottram)
    - Add noop _change_id method (Peter Mottram)

0.006 2014-07-28
  [ CHANGED ]
    - Requires Session::Storage::Secure 0.010 to allow storing objects,
      which is specially relevant to JSON::bool data.

  [ STATISTICS ]
    - code churn: 12 files changed, 423 insertions(+), 477 deletions(-)

0.005 2013-06-05T22:56:23Z America/New_York
  [ FIXED ]
    - Bumped Test::TCP dependency to 1.30 to avoid parallel testing bugs

0.004 2013-05-31T23:40:09Z America/New_York
  [ CHANGED ]
    - Requires Session::Storage::Secure 0.007 for improved resistance to
      timing attacks

0.003 2013-05-28T13:42:39Z America/New_York
  [ FIXED ]
    - Fixed Dancer2 dependency version
    - Fixed tests for new, lazy session creation behavior in Dancer2

0.002 2013-02-22T22:26:31Z America/New_York
  [ CHANGED ]
    - Renamed from Dancer-Session-Factory-Cookie to Dancer2-Session-Cookie
    - Internals updated for new Dancer2 release

0.001 2013-01-24T17:33:57Z America/New_York
  - First release


t/lib/TestApp.pm
package TestApp;

use Dancer2;

get '/no_session_data' => sub {
    return "session not modified";
};

get '/set_session/*' => sub {
    my ($name) = splat;
    session name => $name;
};

get '/read_session' => sub {
    my $name = session('name') || '';
    "name='$name'";
};

get '/destroy_session' => sub {
    my $name = session('name') || '';
    app->destroy_session;
    return "destroyed='$name'";
};

get '/churn_session' => sub {
    app->destroy_session;
    session name => 'damian';
    return "churned";
};

#setting appdir => $tempdir;
#setting( engines => { session => { $engine => $config } } );
#setting( session => $engine );

set(
    show_errors  => 1,
    startup_info => 0,
    environment  => 'production',
);

1;


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.019.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Dancer 2 session storage in secure cookies",
  "dist_author" => [
    "David Golden <dagolden\@cpan.org>"
  ],
  "dist_name" => "Dancer2-Session-Cookie",
  "dist_version" => "0.006",
  "license" => "apache",
  "module_name" => "Dancer2::Session::Cookie",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dancer2" => "0.140001",
    "Dancer2::Core::Role::SessionFactory" => 0,
    "Dancer2::Core::Types" => 0,
    "Moo" => 0,
    "Session::Storage::Secure" => "0.010",
    "perl" => "5.008001",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Dancer2" => "0.140001",
    "File::Spec" => 0,
    "File::Temp" => "0.22",
    "HTTP::Date" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "LWP::UserAgent" => 0,
    "Test::More" => "0.96",
    "Test::TCP" => "1.30",
    "YAML" => 0
  }
);


my %fallback_build_requires = (
  "Dancer2" => "0.140001",
  "File::Spec" => 0,
  "File::Temp" => "0.22",
  "HTTP::Date" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "LWP::UserAgent" => 0,
  "Module::Build" => "0.3601",
  "Test::More" => "0.96",
  "Test::TCP" => "1.30",
  "YAML" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer2/Session/Cookie.pm
use 5.008001;
use strict;
use warnings;

package Dancer2::Session::Cookie;
# ABSTRACT: Dancer 2 session storage in secure cookies
# VERSION

use Session::Storage::Secure 0.011 ();

use Moo;
use Dancer2::Core::Types;

#--------------------------------------------------------------------------#
# Attributes
#--------------------------------------------------------------------------#

=attr secret_key (required)

This is used to secure the cookies.  Encryption keys and message authentication
keys are derived from this using one-way functions.  Changing it will
invalidate all sessions.

=cut

has secret_key => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

=attr default_duration

Number of seconds for which the session may be considered valid.  If
C<cookie_duration> is not set as part of the session configuration,
this is used instead to expire the session after
a period of time, regardless of the length of the browser session.  It is
unset by default, meaning that sessions expiration is not capped.

=cut

has default_duration => (
    is        => 'ro',
    isa       => Int,
    predicate => 1,
);

=attr with_request_address

If set to C<true>, the secret key will have the request address
(as provided by C<<$request->address>>)
appended to it. This can help defeat some replay attacks 
(e.g. if the channel is not secure).
But it will also cause session interruption for people on dynamic addresses.

=cut

has with_request_address => (
    is        => 'ro',
    isa       => Bool,
);

has _store => (
    is      => 'lazy',
    isa     => InstanceOf ['Session::Storage::Secure'],
    handles => {
        '_freeze'   => 'encode',
        '_retrieve' => 'decode',
    },
);

before [qw/ _freeze _retrieve /] => sub {
    my $self = shift;
    return unless $self->with_request_address;
    $self->_store->{secret_key} = join '-', $self->secret_key, $self->request->address;
};

sub _build__store {
    my ($self) = @_;
    my %args = (
        secret_key             => $self->secret_key,
        sereal_encoder_options => { snappy => 1, stringify_unknown => 1 },
        sereal_decoder_options => { validate_utf8 => 1 },
    );
    $args{default_duration} = $self->default_duration
      if $self->has_default_duration;
    return Session::Storage::Secure->new(%args);
}

sub _change_id {
    # This is a noop with session cookies.
    return;
}

with 'Dancer2::Core::Role::SessionFactory';

#--------------------------------------------------------------------------#
# Modified SessionFactory methods
#--------------------------------------------------------------------------#

# We don't need to generate an ID.  We'll set it during cookie generation
sub generate_id { '' }

# Cookie generation: serialize the session data into the session ID
# right before the cookie is generated
before 'cookie' => sub {
    my ( $self, %params ) = @_;
    my $session = $params{session};
    return unless ref $session && $session->isa("Dancer2::Core::Session");
    $session->id( $self->_freeze( $session->data, $session->expires ) );
};

#--------------------------------------------------------------------------#
# SessionFactory implementation methods
#--------------------------------------------------------------------------#

# _retrieve handled by _store

# We don't actually flush data; instead we modify cookie generation
sub _flush { return }

# We have nothing to destroy, either; cookie expiration is all that matters
sub _destroy { return }

# There is no way to know about existing sessions when cookies
# are used as the store, so we lie and return an empty list.
sub _sessions { return [] }

1;

=for Pod::Coverage method_names_here
generate_id

=head1 SYNOPSIS

  # In Dancer 2 config.yml file

  session: Cookie
  engines:
    session:
      Cookie:
        secret_key:           your secret passphrase
        default_duration:     604800
        with_request_address: 0

=head1 DESCRIPTION

This module implements a session factory for Dancer 2 that stores session state
within a browser cookie.  Features include:

=over

=item *

Data serialization and compression using L<Sereal>

=item *

Data encryption using AES with a unique derived key per cookie

=item * 

Enforced expiration timestamp (independent of cookie expiration)

=item * 

Cookie integrity protected with a message authentication code (MAC)

=back

See L<Session::Storage::Secure> for implementation details and important
security caveats.

=head1 SEE ALSO

CPAN modules providing cookie session storage (possibly for other frameworks):

=over

=item * 

L<Dancer::Session::Cookie> -- Dancer 1 equivalent to this module


=item * 

L<Catalyst::Plugin::CookiedSession> -- encryption only


=item * 

L<HTTP::CryptoCookie> -- encryption only

=item * 

L<Mojolicious::Sessions> -- MAC only

=item * 

L<Plack::Middleware::Session::Cookie> -- MAC only

=item * 

L<Plack::Middleware::Session::SerializedCookie> -- really just a framework and you provide the guts with callbacks

=item *

L<Dancer2::Core::Role::SessionFactory> -- documentation of the base package, some more attributes to configure the cookie

=back 

=cut

# vim: ts=4 sts=4 sw=4 et:


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Dancer 2 session storage in secure cookies",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dancer2-Session-Cookie",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Dancer2::Session::Cookie",
  "PREREQ_PM" => {
    "Dancer2" => "0.140001",
    "Dancer2::Core::Role::SessionFactory" => 0,
    "Dancer2::Core::Types" => 0,
    "Moo" => 0,
    "Session::Storage::Secure" => "0.010",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer2" => "0.140001",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "HTTP::Cookies" => 0,
    "HTTP::Request::Common" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Plack" => "1.0029",
    "Plack::Test" => 0,
    "Test::More" => "0.96",
    "YAML" => 0,
    "lib" => 0
  },
  "VERSION" => "0.007",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Dancer2" => "0.140001",
  "Dancer2::Core::Role::SessionFactory" => 0,
  "Dancer2::Core::Types" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTTP::Cookies" => 0,
  "HTTP::Request::Common" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moo" => 0,
  "Plack" => "1.0029",
  "Plack::Test" => 0,
  "Session::Storage::Secure" => "0.010",
  "Test::More" => "0.96",
  "YAML" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
